File: src/content/1.8/code/scala/snippet30.scala
implicit val function1Profunctor = new Profunctor[Function1] {
  override def bimap[A, B, C, D]: (A => B) => (C => D) => (B => C) => (A => D) = ab => cd => bc =>
    cd compose bc compose ab

  override def lmap[A, B, C]: (A => B) => (B => C) => (A => C) =
    flip(compose)

  override def rmap[A, B, C]: (B => C) => (A => B) => (A => C) =
    compose
}

File: src/content/1.8/code/ocaml/snippet30.ml
(* Writing the full implementation without deriving any of the functions *)
module ProfunctorArrow : Profunctor
with type ('a, 'b) t = 'a -> 'b = struct
  type ('a, 'b) t = 'a -> 'b

  let ( % ) = Fun.compose
  let dimap ab cd bc = cd % bc % ab
  let lmap f g = g % f
  let rmap f g = f % g
end
